### DISCUSSION TITLE
How to get subset of a Seurat object based on metadata?

### DISCUSSION THREAD

**Discussion by Sunyp-IM** [satijalab/seurat#7082] [Q&A]
I have a Seurat object in which the meta.data include a column name "predicted_cell_type". The values of this column include "0:CD8 T cell", "1:CD4 T cell", "2:spinous cell", etc.

I would like to extract the "0:CD8 T cell" from the object and use the following command:

`subset(skin, subset = skin@meta.data[['predicted_cell_type']] == '0:CD8 T cell')`
but it gives the following error:

```
Error in FetchData.Seurat(object = object, vars = unique(x = expr.char[vars.use]),  : 
  None of the requested variables were found:
```
And if I use

`subset(skin, subset = predicted_cell_type == '0:CD8 T cell')`
It gives

`Error: No cells found`
Do you know what is the correct command to do this?

With many thanks

**Comment by wudustan** [comment]
`subset(x = skin, cells = WhichCells(skin, expression = predicted_cell_type=="0:CD8 T cell"))`

This should work. Let me know if any issues.

**Reply by XiaoyuZhan520** [reply]
Hello, I met similar situation. 

I want to transfer the target column in metadata (such as the 'cell_type' column) and select the targeted cell type (such as 'tumor_cell') by using argparse function. 

Here is an example. 
```
parser <- argparse::ArgumentParser(description = "do all comparison")
parser$add_argument('-I', '--input', help = 'integrated Seurat object in rds format')
parser$add_argument('-C', '--metacol', help = 'subsetting metadata column')
parser$add_argument('-A', '--anno', help = 'subsetting annotation/cell type')
opts <- parser$parse_args()

seurat_obj <- readRDS(opts$input)
subset <- subset(x = seurat_obj, subset = opts$metacol  ==  opts$anno)
```

However, it could not work. Could you give any suggestion? Many thanks in advance!

**Reply by wudustan** [reply]
Hi, I don't have experience of argpase, could you show me a limited output of 'opts$metacol' and 'opts$anno'. My kneejerk thought is that `  ==  opts$anno` will only work if  `opts$anno` is a single value or the same values as `opts$metacol` (in which case the subset is not worth running).

**Reply by XiaoyuZhan520** [reply]
Sure, I am glad to provide you more information. 

The R script
```
library(Seurat)
parser <- argparse::ArgumentParser(description = "do all comparison")
parser$add_argument('-I', '--input', help = 'integrated Seurat object in rds format')
parser$add_argument('-C', '--metacol', help = 'subsetting metadata column')
parser$add_argument('-A', '--anno', help = 'subsetting annotation/cell type')
opts <- parser$parse_args()
class(opts$metacol)
class(opts$anno)
seurat_obj <- readRDS(opts$input)
subset <- subset(x = seurat_obj, subset = opts$metacol  ==  opts$anno)

```
The command for running the R script
`Rscript test.R -I test.rds -C anno3 -A Tumor
`
The output of R information
```
[1] "character"
[1] "character"
Error in `FetchData()`:
! None of the requested variables were found: 
Backtrace:
    ▆
 1. ├─base::subset(x = seurat_obj, subset = opts$metacol == opts$anno)
 2. └─SeuratObject:::subset.Seurat(...)
 3.   ├─SeuratObject::WhichCells(...)
 4.   └─SeuratObject:::WhichCells.Seurat(...)
 5.     ├─SeuratObject::FetchData(...)
 6.     └─SeuratObject:::FetchData.Seurat(...)
 7.       └─rlang::abort(...)
Execution halted

```
After a few try, I found opts$anno could be accepted by Seurat, but not the opts$metacol. 
When I write, it works. 
`subset <- subset(x = seurat_obj, subset = anno3 ==  opts$anno)
`

I would appreciate if you could give me any suggestion. Thanks!

**Comment by lzmcboy** [comment]
I usually use the script below:

```
Idents(skin) <- "predicted_cell_type"  
cd8_subset <- subset(skin, ident = "0:CD8 T cell")
```

**Comment by julia-aguade** [comment]
would it be possible to do something similar but include 2 or groups from the same metadata column in the subset? For instance I have metadata "orig.ident" which is the tissue type and has 6 possible options (liver, bone_marrow, spleen...). How can I make a subset including only the cells with liver and bone_marrow in the orig.ident? I tried orig.ident == c("liver","bone_marrow") but got an error.

thanks!

> `subset(x = skin, cells = WhichCells(skin, expression = predicted_cell_type=="0:CD8 T cell"))`
> 
> This should work. Let me know if any issues.

**Reply by samuel-marsh** [reply]
Hi,

The issue you are having is R fundamental. `==` vs `%in%`.  See here for some more info: https://stackoverflow.com/questions/42637099/difference-between-the-and-in-operators-in-r

Best,
Sam

**Comment by ACMGlueck** [comment]
I am having similar problems as described above, and the above suggestions don't work for me.  Here is a very simple subsetting example, and even that doesn't work for me:
```
mySeuratObject[1:3, 1:10]
RNA assay doesn't leave any cells, so it is removed
Error in subset.Seurat(x = x, features = i, cells = j, ...) : 
  No cells left in the default assay, please change the default assay
In addition: Warning message:
Cannot find cells provided
```
In my understanding, [1:3, 1:10] should be as simple as possible to create a subset, and it works with `pbmc_small[1:3, 1:10]`. mySeuratObject has more than 3 genes and 10 cells, btw.
Any insights on what could be my problem will be very welcome.

**Reply by idibaps001** [reply]
I have the same problem

**Reply by samuel-marsh** [reply]
@ACMGlueck @idibaps001,

Can you provide reproducible example with dataset from Seurat data?  I cannot replicate the issue in my testing.  Have you also tried using `subset` and setting `cells` and `features` parameter?  Does that result in the same issue?

Best,
Sam

**Reply by idibaps001** [reply]
https://www.biostars.org/p/9596406/     Answered by Aurel.   "That's likely because the rownames in your seurat metadata do not match the colnames of your count matrix".      I checked this from my dataset, he is correct. My dataset is downloaded from a public platform, and it is not a common seurat object.
